From 490d9808e9adf0d12e34032ec6b7c85578275ab2 Mon Sep 17 00:00:00 2001
From: Vasilis Liaskovitis <vliaskovitis@suse.com>
Date: Wed, 19 Jul 2017 16:51:29 +0200
Subject: xen: hold lock_device_hotplug throughout vcpu hotplug operations
Patch-mainline: Never, SUSE specific
References: bsc#1042422

Holding the lock avoids crashes when issuing mulitple hot-add
/ hot-remove operations to the same vCPU(s) back to back.

Signed-off-by: Vasilis Liaskovitis <vliaskovitis@suse.com>
Acked-by: Juergen Gross <jgross@suse.com>
---
 drivers/xen/cpu_hotplug.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/drivers/xen/cpu_hotplug.c b/drivers/xen/cpu_hotplug.c
index 5676aef..ce626a5 100644
--- a/drivers/xen/cpu_hotplug.c
+++ b/drivers/xen/cpu_hotplug.c
@@ -10,23 +10,24 @@
 
 static void enable_hotplug_cpu(int cpu)
 {
+	lock_device_hotplug();
 	if (!cpu_present(cpu))
 		xen_arch_register_cpu(cpu);
 
 	set_cpu_present(cpu, true);
+	unlock_device_hotplug();
 }
 
 static void disable_hotplug_cpu(int cpu)
 {
-	if (cpu_online(cpu)) {
-		lock_device_hotplug();
+	lock_device_hotplug();
+	if (cpu_online(cpu))
 		device_offline(get_cpu_device(cpu));
-		unlock_device_hotplug();
-	}
 	if (cpu_present(cpu))
 		xen_arch_unregister_cpu(cpu);
 
 	set_cpu_present(cpu, false);
+	unlock_device_hotplug();
 }
 
 static int vcpu_online(unsigned int cpu)
-- 
1.7.12.4

